### ISSUE TITLE
Adding custom method breaks gradient flow

### ISSUE THREAD

**Issue by ChintanTrivedi** [vturrisi/solo-learn#231]
I am trying to modify the [SimCLR training_step method](https://github.com/vturrisi/solo-learn/blob/main/solo/methods/simclr.py#L91) to create a new loss formulation for my custom unlabelled dataset. However, this is leading to the backbone weights not being updated during training. Could you please guide me how to find out the breaking point of gradient flow in the graph? I have not touched any other part of the framework except this training step function.

Here's my modification of the SimCLR method:-
```
 def training_step(self, batch: Sequence[Any], batch_idx: int) -> torch.Tensor:
       _, X, targets = batch

        # use only one augmentation from the batch
        batch_combined_images = X[0]
        batch_size = batch_combined_images.shape[0]
        img_crop_size = batch_combined_images.shape[2]
        num_images_per_item = int(batch_combined_images.shape[3] / img_crop_size)

        # split the image into 7 items (7 images of size 3x224x224 each)
        X_split = torch.empty((num_images_per_item * batch_size, 3, img_crop_size, img_crop_size),
                              dtype=torch.float32).to("cuda")

        # prepare custom labels/targets for the separated images
        con_label = []
        con_label_ctr = 0
        for item_idx in range(batch_size):
            for image_idx in range(num_images_per_item):
                image = batch_combined_images[item_idx, :, :,
                        image_idx * img_crop_size:image_idx * img_crop_size + img_crop_size]
                X_split[item_idx * num_images_per_item + image_idx] = image

                if image_idx > 3:
                    con_label_ctr += 1
                con_label.append(con_label_ctr)
            con_label_ctr += 1
        targets = torch.as_tensor(con_label, dtype=torch.int64).to("cuda")

        out = super().training_step([targets, [X_split, X_split], targets], batch_idx)

        feats = out["feats"][0]
        z = self.projector(feats)

        nce_loss = simclr_loss_func(
            z,
            indexes=targets,
            temperature=self.temperature,
        )

        self.log("train_nce_loss", nce_loss, on_epoch=True, sync_dist=True)

        return nce_loss
```

PS: I realize the redundancy in the augmentation pipeline, I will address it once I get the training working.

**Comment by vturrisi** [comment]
Hey,
If X_split has gradients (I would check it), I don't see any clear issues. For sure it's something related to how you create your data.

**Comment by vturrisi** [comment]
You are passing X_split twice? You need different augmentations, otherwise, the model will produce the same output for both views and I think the loss is just 0

**Comment by ChintanTrivedi** [comment]
I found the issue, and it is related to the data I'm using and not the code itself. 

Since the loss function is implemented to always have positives for each image (two augmented views of the same input), the `pos_mask` for positive pairs will never have zeros in it. However, since my modified batch has certain images that only act as negative pairs to the anchor image, they themselves do not have any other positive pairings in my batch. This is the reason why loss/gradients were not computed properly and hence the weights were not updating.

I fixed it by modifying the simclr loss to exclude those images from numerator that do not have positives:-
`loss = -(torch.mean(torch.log(pos[pos != 0] / (pos[pos != 0] + neg[pos != 0]))))`

**Comment by vturrisi** [comment]
Ok, so the loss was going to NaN, right? I think a better fix would be something like:
```
  ....
  neg_mask = indexes.t() != gathered_indexes

  # filter when there's not a single positive for some instance
  has_pos = pos_mask.sum(dim=1) > 0
  sim = sim[has_pos]
  pos_mask = pos_mask[has_pos]
  neg_mask = neg_mask[has_pos]
  
  pos = torch.sum(sim * pos_mask, 1)
  ...
```
Anyway, glad that you manage to solve it.
